src/controller.js

```diff
@@ -0,0 +1,601 @@
+    }));
+
+    let {bucket, key, tags} = opts;
+
+    let requestBody = this.__generateTagSetBody(tags);
+
+    let signedRequest = aws4.sign({
+      service: 's3',
+      region: this.region,
+      method: 'PUT',
+      protocol: this.s3protocol,
+      hostname: this.__s3hostname(bucket),
+      path: `/${key}?tagging=`,
+      body: requestBody,
+    });
+
+    let response = await this.runner({
+      req: this.__serializeRequest(signedRequest),
+      body: requestBody,
+    });
+
+
+    parseS3Response(response.body);
+
+    if (response.statusCode !== 200) {
+      throw new Error('Could not tag object');
+    }
+  }
+
+  /**
+   * Mark a multipart upload as completed. 
+   *
+   * http://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadComplete.html
+   */
+  async completeMultipartUpload(opts) {
+    opts = runSchema(opts, Joi.object({
+      bucket: schemas.bucket.required(),
+      key: schemas.key.required(),
+      etags: schemas.etags.required(),
+      tags: schemas.tags,
+      uploadId: schemas.uploadId.required(),
+    }).optionalKeys('tags'));
+
+    let {bucket, key, uploadId, etags, tags} = opts;
+
+    // I'm not sure why, but for some reason the AWS4 library generates the
+    // incorrect SHA256 for *this* and only *this* body.  I have to calculate
+    // the body sha256 myself, or else what happens is that the requestBody
+    // string is hashed with its value and an extra newline character, but then
+    // the data written does not have that newline, and never did.  I suspect
+    // that there's something broken in the aws4 library here
+    let requestBody = this.__generateCompleteUploadBody(etags);
+    let requestBodySha256 = crypto.createHash('sha256').update(requestBody);
+
+    let signedRequest = aws4.sign({
+      service: 's3',
+      region: this.region,
+      method: 'POST',
+      protocol: this.s3protocol,
+      hostname: this.__s3hostname(bucket),
+      path: `/${key}?uploadId=${uploadId}`,
+      headers: {
+        'X-Amz-Content-Sha256': requestBodySha256.digest('hex'),
+        'Content-Length': requestBody.length,
+      },
+    });
+
+    let response = await this.runner({
+      req: this.__serializeRequest(signedRequest),
+      body: requestBody,
+    });
+
+    let parsedResponse = parseS3Response(response.body);
+
+    let multipartEtag = this.__getMultipartEtag(parsedResponse, bucket, key);
+
+    if (response.statusCode !== 200) {
+      throw new Error('Could not complete a multipart upload');
+    }
+
+    if (tags) {
+      await this.__tagObject(opts);
+    }
+
+    return multipartEtag;
+  }
+
+  // http://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadAbort.html
+  async abortMultipartUpload(opts) {
+   opts = runSchema(opts, Joi.object({
+     bucket: schemas.bucket.required(),
+     key: schemas.key.required(),
+     uploadId: schemas.uploadId.required(),
+   }));
+
+    let {bucket, key, uploadId} = opts;
+    let signedRequest = aws4.sign({
```

Comment:
There's no reason to use aws-sdk.  We need to generate a signature for all requests.  This is doing it with the aws4 library instead of the aws-sdk library, but it's doing the exact same thing under the hood.